src/libxml/node_manip.cxx

```diff
@@ -45,21 +45,45 @@
 namespace
 {
 
+// Check if two XML namespaces are the same. Both pointers must be non-null.
+bool are_same(xmlNsPtr ns1, xmlNsPtr ns2)
+{
+    if ( ns1->type != ns2->type )
+        return false;
+
+    if ( xmlStrcmp(ns1->href, ns2->href) != 0 )
+        return false;
+
+    if ( ns1->prefix )
+    {
+        if ( !ns2->prefix || xmlStrcmp(ns1->prefix, ns2->prefix) != 0 )
+            return false;
+    }
+    else
+    {
+        if ( ns2->prefix )
+            return false;
+    }
+
+    return true;
+}
+
 // Helper: sets the given namespace for this node and all of its children,
 // recursively, but without overriding any existing namespaces.
 void set_default_ns_recursively(xmlNodePtr node, xmlNsPtr ns)
 {
-    if ( node->ns )
+    if ( node->ns && !are_same(node->ns, ns) )
```

Comment:
No, it is really needed, i.e. it can't be just removed because if a child has a pointer to the same namespace as is about to be freed in `copy_node_under_parent()`, we must replace it, otherwise it would end up with a dangling pointer. But you're right that we can do this without `are_same()` and just compare pointers instead, I'll do this in the updated version of this PR.